File: packages/nodejs/agenda-cobranca/src/signer.ts
import { createHash, createHmac, randomUUID, timingSafeEqual } from "node:crypto";
import { ConfigurationError } from "./errors.js";
import type { SignedHeaders } from "./types.js";

export type BodyInput = string | Uint8Array | null | undefined;

export function bodyBytes(body: BodyInput): Buffer {
  if (body == null) {
    return Buffer.alloc(0);
  }
  if (typeof body === "string") {
    return Buffer.from(body, "utf8");
  }
  return Buffer.from(body);
}

export function hashBody(body: BodyInput = ""): string {
  return createHash("sha256").update(bodyBytes(body)).digest("hex");
}

export function canonicalString(
  method: string,
  path: string,
  timestamp: string | number,
  nonce: string,
  body: BodyInput = "",
): string {
  return [
    method.toString().toUpperCase(),
    path.toString(),
    timestamp.toString(),
    nonce.toString(),
    hashBody(body),
  ].join("\n");
}

export class Signer {
  private readonly clientSecret: string;
  private readonly clock: () => number;
  private readonly nonceGenerator: () => string;

  constructor(
    clientSecret: string,
    clock: () => number = () => Math.floor(Date.now() / 1000),
    nonceGenerator: () => string = () => randomUUID(),
  ) {
    if (!clientSecret || clientSecret.trim() === "") {
      throw new ConfigurationError("client_secret e obrigatorio");
    }
    this.clientSecret = clientSecret;
    this.clock = clock;
    this.nonceGenerator = nonceGenerator;
  }

  signatureFor(
    method: string,
    path: string,
    timestamp: string | number,
    nonce: string,
    body: BodyInput = "",
  ): string {
    const canonical = canonicalString(method, path, timestamp, nonce, body);
    return createHmac("sha256", this.clientSecret).update(canonical, "utf8").digest("hex");
  }

  sign(method: string, path: string, body: BodyInput = ""): SignedHeaders {
    const timestamp = this.clock().toString();
    const nonce = this.nonceGenerator().toString();
    return {
      timestamp,
      nonce,
      signature: this.signatureFor(method, path, timestamp, nonce, body),
      bodyHash: hashBody(body),
    };
  }

  validSignature(
    method: string,
    path: string,
    timestamp: string,
    nonce: string,
    signature: string,
    body: BodyInput = "",
  ): boolean {
    const expected = this.signatureFor(method, path, timestamp, nonce, body);
    const left = Buffer.from(expected, "utf8");
    const right = Buffer.from(signature ?? "", "utf8");
    if (left.length !== right.length) {
      return false;
    }
    return timingSafeEqual(left, right);
  }
}

File: packages/ruby/agenda_cobranca/lib/agenda_cobranca/security/signer.rb
# frozen_string_literal: true

require "digest"
require "openssl"
require "securerandom"

module AgendaCobranca
  module Security
    class Signer
      def initialize(client_secret, clock: -> { Time.now.to_i }, nonce_generator: -> { SecureRandom.uuid })
        raise ConfigurationError, "client_secret e obrigatorio" if client_secret.nil? || client_secret.to_s.empty?

        @client_secret = client_secret.to_s
        @clock = clock
        @nonce_generator = nonce_generator
      end

      def sign(method:, path:, body: "")
        timestamp = @clock.call.to_i.to_s
        nonce = @nonce_generator.call.to_s
        signature = signature_for(method: method, path: path, timestamp: timestamp, nonce: nonce, body: body)

        {
          timestamp: timestamp,
          nonce: nonce,
          signature: signature,
          body_hash: hash_corpo(body)
        }
      end

      def signature_for(method:, path:, timestamp:, nonce:, body: "")
        canonical = canonical_string(
          method: method,
          path: path,
          timestamp: timestamp,
          nonce: nonce,
          body: body
        )
        hmac_hex(canonical)
      end

      def canonical_string(method:, path:, timestamp:, nonce:, body: "")
        [
          method.to_s.upcase,
          path.to_s,
          timestamp.to_s,
          nonce.to_s,
          hash_corpo(body)
        ].join("\n")
      end

      def hash_corpo(body)
        Digest::SHA256.hexdigest(corpo_bytes(body))
      end

      def valid_signature?(method:, path:, timestamp:, nonce:, body:, signature:)
        esperado = signature_for(
          method: method,
          path: path,
          timestamp: timestamp,
          nonce: nonce,
          body: body
        )
        comparar_seguro(esperado, signature.to_s)
      end

      private

      def hmac_hex(canonical)
        OpenSSL::HMAC.hexdigest("SHA256", @client_secret, canonical)
      end

      def corpo_bytes(body)
        return "" if body.nil?

        body.is_a?(String) ? body : body.to_s
      end

      def comparar_seguro(esperado, recebido)
        return false if esperado.bytesize != recebido.bytesize

        OpenSSL.fixed_length_secure_compare(esperado, recebido)
      end
    end
  end
end
